lib/llm/src/protocols/openai/validate.rs

```diff
@@ -0,0 +1,529 @@
+pub const MAX_TOP_P: f32 = 1.0;
+/// Allowed range of values for OpenAI's `top_p` sampling option
+pub const TOP_P_RANGE: (f32, f32) = (MIN_TOP_P, MAX_TOP_P);
+
+/// Minimum allowed value for OpenAI's `frequency_penalty` sampling option
+pub const MIN_FREQUENCY_PENALTY: f32 = -2.0;
+/// Maximum allowed value for OpenAI's `frequency_penalty` sampling option
+pub const MAX_FREQUENCY_PENALTY: f32 = 2.0;
+/// Allowed range of values for OpenAI's `frequency_penalty` sampling option
+pub const FREQUENCY_PENALTY_RANGE: (f32, f32) = (MIN_FREQUENCY_PENALTY, MAX_FREQUENCY_PENALTY);
+
+/// Minimum allowed value for OpenAI's `presence_penalty` sampling option
+pub const MIN_PRESENCE_PENALTY: f32 = -2.0;
+/// Maximum allowed value for OpenAI's `presence_penalty` sampling option
+pub const MAX_PRESENCE_PENALTY: f32 = 2.0;
+/// Allowed range of values for OpenAI's `presence_penalty` sampling option
+pub const PRESENCE_PENALTY_RANGE: (f32, f32) = (MIN_PRESENCE_PENALTY, MAX_PRESENCE_PENALTY);
+
+/// Maximum allowed value for `top_logprobs`
+pub const MIN_TOP_LOGPROBS: u8 = 0;
+/// Maximum allowed value for `top_logprobs`
+pub const MAX_TOP_LOGPROBS: u8 = 20;
+
+/// Minimum allowed value for `logprobs` in completion requests
+pub const MIN_LOGPROBS: u8 = 0;
+/// Maximum allowed value for `logprobs` in completion requests
+pub const MAX_LOGPROBS: u8 = 5;
+
+/// Minimum allowed value for `n` (number of choices)
+pub const MIN_N: u8 = 1;
+/// Maximum allowed value for `n` (number of choices)
+pub const MAX_N: u8 = 128;
+
+/// Minimum allowed value for OpenAI's `logit_bias` values
+pub const MIN_LOGIT_BIAS: f32 = -100.0;
+/// Maximum allowed value for OpenAI's `logit_bias` values
+pub const MAX_LOGIT_BIAS: f32 = 100.0;
+
+/// Minimum allowed value for `best_of`
+pub const MIN_BEST_OF: u8 = 0;
+/// Maximum allowed value for `best_of`
+pub const MAX_BEST_OF: u8 = 20;
+
+/// Maximum allowed number of stop sequences
+pub const MAX_STOP_SEQUENCES: usize = 4;
+/// Maximum allowed number of tools
+pub const MAX_TOOLS: usize = 128;
+/// Maximum allowed number of metadata key-value pairs
+pub const MAX_METADATA_PAIRS: usize = 16;
+/// Maximum allowed length for metadata keys
+pub const MAX_METADATA_KEY_LENGTH: usize = 64;
+/// Maximum allowed length for metadata values
+pub const MAX_METADATA_VALUE_LENGTH: usize = 512;
+/// Maximum allowed length for function names
+pub const MAX_FUNCTION_NAME_LENGTH: usize = 64;
+/// Maximum allowed value for Prompt IntegerArray elements
+pub const MAX_PROMPT_TOKEN_ID: u16 = 50256;
+
+//
+// Shared Fields
+//
+
+/// Validates the temperature parameter
+pub fn validate_temperature(temperature: Option<f32>) -> Result<(), anyhow::Error> {
+    if let Some(temp) = temperature {
+        if !(MIN_TEMPERATURE..=MAX_TEMPERATURE).contains(&temp) {
+            anyhow::bail!(
+                "Temperature must be between {} and {}, got {}",
+                MIN_TEMPERATURE,
+                MAX_TEMPERATURE,
+                temp
```

Comment:
These days you can do `"Temperature must be between {MIN_TEMPERATURE} and {MAX_TEMPERATURE}, got {temp}"`